buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java

```diff
@@ -466,48 +472,67 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
-            lineCutoff = lines.count();
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
-        final Map<String, Long> errorsAndWarnings;
-        try(Stream<String> lines = Files.lines(from, StandardCharsets.UTF_8)) {
-            errorsAndWarnings = lines
-                .filter(line -> line.contains("ERROR") || line.contains("WARN"))
-                .map(this::normalizeLogLine)
-                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
-        lineCutoff -= 50;
+        final Map<String, Integer> errorsAndWarnings = new LinkedHashMap<>();
+        LinkedList<String> ring = new LinkedList<>();
         try (LineNumberReader reader = new LineNumberReader(Files.newBufferedReader(from))) {
-            String line = null;
-            do {
-                line = reader.readLine();
-                if (line == null) {
-                    continue;
-                }
-                String normalizedLine = normalizeLogLine(line);
-                if (errorsAndWarnings.keySet().contains(normalizedLine)) {
-                    logger.lifecycle("  " + line);
-                    if (errorsAndWarnings.get(normalizedLine) != 1) {
-                        logger.lifecycle("    ↑ repeated another " + errorsAndWarnings.get(normalizedLine) + " times ↑");
+            for (String line = reader.readLine(); line != null ; line = reader.readLine()) {
+                final String lineToAdd;
+                if (line.startsWith("[")) {
+                    // check to see if the previous message (possibly combined from multiple lines) was an error or
+                    // warning as we want to show all of them
+                    if (ring.isEmpty() == false) {
+                        String previousMessage = normalizeLogLine(ring.getLast());
+                        if (MESSAGES_WE_DONT_CARE_ABOUT.stream().anyMatch(match -> previousMessage.contains(match))) {
+                            ring.removeLast();
+                        } else {
+                            if (previousMessage.contains("ERROR") || previousMessage.contains("WARN")) {
+                                errorsAndWarnings.put(
+                                    previousMessage,
+                                    errorsAndWarnings.getOrDefault(previousMessage, 0) + 1
+                                );
+                            }
+                        }
+                    }
+                    lineToAdd = line;
+                } else {
+                    // We combine multi line log messages to make sure we never break exceptions apart
+                    if (ring.isEmpty()) {
+                        lineToAdd = line;
+                    } else {
+                        lineToAdd = ring.removeLast() + "\n" + line;
```

Comment:
This is going to be very inefficient, creating tons of strings whenever we encounter an exception. I wonder if we really need the de-duplication (as @mark-vieira suggested earlier). The algorithm we have now in RestIntegTestTask doesn't require any additional memory. It does mean we lean towards the beginning of the log instead of the end, but I think this is ok.

Reply:
@rjernst I do think the de-duplication is useful, I got inspired to implement it by looking at test failures when converting stuff to test clusters, and otherwise I think we tend  to have repeating warnings in the logs when something happens. Makes it easy to tell if something else was also going on that would be hard to fin between the repeating lines. 
I'm not sure it the inefficiency will be noticeable for anything but huge logs with lots of exceptions, if we're concerned abut that we could tail  e.x. the last 10K of the log

Reply:
> I'm not sure it the inefficiency will be noticeable for anything but huge logs with lots of exceptions

This is exactly the case when we don't the build to fail on us. I don't think we should add anything blatantly unperformant when large logs occur since they do happen.

I think we should keep the original single pass that we have in RestIntegTestTask. If we want to have deduplication, it should be best effort, using a cache of the last N messages seen.